add train test split

stereomatchingkiss opened the pull request:
Add function to split train and test data

zoq commented:
hanks for the pull request I think, this is the first step to get the GSOC 'Dataset and experimentation tools' project started.

I took a quick look over the pull request and have some comments:

- Instead of returning a tuple that contains the train and test data, I would go with references, to be more consistent with the other mlpack code that has functions like this.

``TrainTestSplit(const MatType& input, MatType& train, MatType& test)``

- You can create a random vector of indexes using armadillo instead of using an std::vector.

`'arma::Col<size_t> order = arma::shuffle(arma::linspace<arma::Col<size_t> >(0, (details::extractSize(input) - 1), details::extractSize(input)));``

 I guess this is just a preference of mine, because your code is absolutely fine. I've seen this code fragment a couple of times, so I guess I'm biased.

 - Please use camel casing for all names. Capitalize class and method names.

Another thing that crossed my mind is, can we modify the code in a why that it also works with data matrices with more than one dimension. Imagine a dataset of images with more than one channel per image.

stereomatchingkiss commented:
>I would go with references, to be more consistent with the other mlpack code that has functions like this.

I suggest we add another overload to accept the reference and keep the old api, because 
In this case, there would be 6 parameters need to pass into the api

TrainTestSplit(const MatType& input, MatType& train, MatType& test, Row<size_t> &trainLabel,
Row<size_t> &testLabel, double testRatio, unsigned int seed = 0);

>You can create a random vector of indexes using armadillo instead of using an std::vector.

Thanks, arma version looks neat and shorter

>works with data matrices with more than one dimension. Imagine a dataset of images with more than one channel per image.

I beg of your pardon, could you show me more details?Do you mean

a : Treat arma::Mat as a whole image(1~n channels), arma::Cube as a container of those images

I think this is not a problem, the overload of arma::Cube already solve this problem. We could store pixels of multi channels image into arma::Mat, store arma::Mat(from one channel to n channels) into arma::Cube, then split them.

b : You want to split arma::Mat of n into different channels?

If it is case b, different libraries may have different way to store their pixels, I think it is better to let the users preprocess their data first

or another cases?

zoq commented:
I guess a simple example would make things more clear. Imagine you have a dataset with a bunch of colour images (e.g. CIFAR-10), so more than one channel per image or dataset, something like that:

<p align="center">
<img src="https://urgs.org/marcus/media/we_need_to_go_deeper.png" alt="we need to go deeper color">
</p>

So, I guess I would store that image in a cube with 3 slices one slice per channel. But now I would like to store 10 images in the same data structure that would result in a cube with 10 * 3 slices.

As far as I can see is that your code works with data sets that contains one channel per image, something like that would be totally fine:

<p align="center">
<img src="https://urgs.org/marcus/media/we_need_to_go_deeper_gray.png" alt="we need to go deeper gray">
</p>

As I said I'm not sure we should handle such datasets with the code. I guess, we have to specify the number of channels/slices per dataset somehow to make it work.

I hope this makes some sense.

stereomatchingkiss commented:
I think I got your point, after looking at the codes of convolution_layer, what you want to do is something like solution A.

Solution A : split three channels to different arma::mat

```
cv::Mat img = cv::imread("00002.jpg");
std::vector<cv::Mat> channels;
cv::split(img, channels);
//........convert channels to arma::mat
arma::cube cubes;
join_slices(cubes, armaCh[0]);
join_slices(cubes, armaCh[1]);
join_slices(cubes, armaCh[2]);
```

Yeah, I think we should support this case, I prefer write a class to do the task since this would be a lot of parameters need to specify. Please correct me if I misunderstand your points.

stereomatchingkiss commented:
One question, when I try to provide the support of multi slice(depth), I find out one problem--what should the labels organized?

Assume we have 2 images with 3 channels,the labels should be
Solution 1 : 
0, 1 
or

Solution 2 : 
 0, 0, 0, 1, 1, 1

According to current source codes, the train class always take one slice, so I should take solution 2, what is your opinions?Thanks

stereomatchingkiss commented:
I take a quick look at the codes of ann, it changed a lot. Old Trainer cannot work anymore, do you plan to customize every optimizer and give up the Trainer class?  Could you tell me the reasons and the pros/cons of the new designs(I hope I could learn something from it :) ), thanks.

zoq commented:
Basically we thought it would be a nice idea to make the ANN code work with the mlpack optimizers. But also the other way around, once we have rewritten the ann optimizer (I have already rewritten the RMSprop and Adam optimizer), we could use them for, e.g. logistic regression. Although solves this the serialization problem of all layers that previously hold an optimizer. At the same time, we also removed the weight initialization rule from each layer. Because, its way easier to initialize the weights, e.g. using a gaussian distribution with zero mean and unit variance. Actually you could do the same thing with the previous layer architecture, by creating an initial weight matrix for the whole network and then initialize the weights using that matrix. Another reason for the transition, was that we could use the ANN code the same way as we use, e.g. the logistic regression or the naive bayes classifier. Obviously there is a drawback, basically you can't specify another optimizer for each layer.

So, yes, the plan is to delete all ANN optimizers and the general trainer class in favour of a unified interface and the ability to the existing mlpack optimizers.

Keep in mind that the optimizer is independent of the underlying data structure and model. The optimizer tells the underlying function, e.g. the convolutional neural network which set to choose from the complete dataset and the function/model does the rest. This works because the optimizer calls the following function to evaluate a set:
```
function.Evaluate(weight, i);
```
The interesting parameter is ```'i'```, which tells the function to evaluate the i'th set of the complete dataset. Btw. the dataset is hold by the underlying function and not the optimizer, which is great because the function e.g. the convolutional neural network knows best how to work with the dataset. So, we could tell the network, to interpret that one set of the dataset is actually a cube with 3 slices. And don't have to modify the optimizer or trainer.

If you have any questions about the code, do not hesitate to ask me.

Actually I have a preference for solution 1. As I pointed out above the model holds the data and knows the best how to handle it. We could add a parameter to the CNN class that tells the network how to interpret the data. E.g. the number of slices that build a set of the dataset. Something like that 

```CNN(input, labels, dataSize) { ... }```

The evaluation function of the convolutional neural network which is called by the optimizer can then choose the right number of slices.


I was only bringing this up because, I was thinking if we can use your TrainTestSplit function to shuffle and split a dataset that is basically a cube, but that contains data, which is formed by more than one slice.
```
arma::cube image1 = arma::cube(23, 23, 3);
arma::cube image2 = arma::cube(23, 23, 3);

arma::cube data = arma::cube(23, 23, 6);
data.slices(0, 2) = image1;
data.slices(3, 5) = image2;
```

We can't just shuffle the dataset (data) because one set of the has 3 slices.

> ps : the images are funny:)

I was planning to use the images, for the GoogLeNet GSoC idea, but I figured that this doesn't make sense at all, so I took the opportunity :)

stereomatchingkiss commented:
>We can't just shuffle the dataset (data) because one set of the has 3 slices.

It can be done, but need some change.First part is the ExtractData

 ```
 template<typename T>
  void ExtractData(arma::Cube<T> const &input, arma::Cube<T> &output,
                   size_t inputIndex, size_t outputIndex) const
  {
    outputIndex *= slice;
    inputIndex *= slice;
    output.slices(outputIndex, outputIndex + slice - 1) =
        input.slices(inputIndex, inputIndex + slice - 1);
  }
```

Second part is augmentation the order of col(if you take solution 2)

```
arma::Col<size_t> augmentOrder(order.size()*slice);
for(size_t i = 0; i != order.size(); ++i){
    for(size_t j = 0; j != slice; ++j){
       augmentOrder(i*slice + j) = order(i);
    }
}
```

>Actually I have a preference for solution 1
Then I would take the path of solution 1, this one could save some memory and maybe more intuitive for users.

rcurtin commented:
The way `arma::Cube` holds data is "slice-major": the data in an individual slice is all contiguous.  Each slice is column-major.  So if an `arma::Cube` is going to be used to hold data, then it would probably be best if each slice represented an individual point in the dataset.

But a problem with this approach is that now each point is accessed with `.slice(i)` instead of `.col(i)`.  This means that a user can't use an `arma::Cube` in this way and pass it as a `MatType` parameter to, say, logistic regression.  i.e. this is not possible with each point as a slice: `LogisticRegression<arma::Cube<double>>`.

In previous situations where images are dealt with or other multi-dimensional structures, they are basically "vectorized": a 256x256 image is just treated as a 65536-dimensional vector.  The same thing could be done with channels of an image: a 256x256x3 image becomes a 196608-dimensional vector.  And in that case, a column is a single point, like in the rest of mlpack.  For input into a neural network, I think that any image is vectorized in this way anyway (maybe I am incorrect here?), so personally I think it makes the most sense to just work with a vectorized representation the whole time.

If we decide to work with arbitrary types of data in higher-dimensional interpretations instead of vectorizing it, we'll have to rethink some of the core abstractions of mlpack, because "one column is one point" will no longer apply.  I'm not saying that we shouldn't make that change, just that we'll have to think through it pretty clearly so we can continue to present a unified interface.

stereomatchingkiss commented:
>a 256x256x3 image becomes a 196608-dimensional vector. And in that case, a column is a single point, like in the rest of mlpack.

This is the original solution I provided, current implementation can generate same result too if you specify the slice == 1

zoq commented:
I agree, using a vector representation for an image or a multi-channel image, is probably the best solution. So I guess, the question is, should the function to split train and test data support cubes as it does right now or should we remove that support? If we decide to support cubes, should we also support that a set could be more than one slice?

stereomatchingkiss commented:
>If we decide to support cubes, should we also support that a set could be more than one slice?

Do you mean you want to provide a function to let the users copy n channels image into arma::mat?

Example : 
```
cv::Mat img = cv::imread("ttt.jpg");
//img has 3 channels
cv::arma::mat armaImg;

CopyToArma(img.ptr<uchar>(0),  armaImg, img.rows, img.cols, img.channels);

//layout of armaImg will be
//b b b b b b b b.....
//g g g g g g g g.....
//r  r  r  r  r  r  r  r..... 
```

Different libraries has different way to store their data, do we need to take care of that too?

rcurtin commented:
Personally I think that we should avoid supporting Cubes, and have users represent their data always as an arma::Mat object.  If they want they can actually represent their data in their program as a Cube, but when they use mlpack they should represent it as a Mat.  This is pretty easy actually:

```
arma::Cube cube; // this is filled with data, each slice is a point
arma::Mat matrixRepresentation(cube.memptr(), cube.n_slices, cube.n_rows * cube.n_cols, false); // don't copy memory; this is an alias
```

Maybe there is some nicer Armadillo function to do that; I didn't look.

stereomatchingkiss commented:
>Personally I think that we should avoid supporting Cubes

I think Cube is essential for ann module, it would be much harder to store and extract the data without cube.

rcurtin commented:
Do you mean that it's essential internally for the ANN code, or that the user should pass in their data as an `arma::Cube`?  I was only referring to the latter---that users should represent their data as `arma::mat`s.

stereomatchingkiss commented:
>Do you mean that it's essential internally for the ANN code, or that the user should pass in their data as an arma::Cube?

I think both? At least for cnn, it would be easier to deal with the images if we can treat the cube as the container of  images(with n channels). Exactly, it do not need to be cube, it could be std::vector, we just need a container to store those images, but using arma cube may fit mlpack better.

zoq commented:
We can also use ```arma::Mat``` to store images as @rcurtin suggested. I guess the only layer we have to modify is the convolution layer. I'm working on another convolution method that should be faster as the already implemented methods for small kernels. The method uses ```arma::Mat``` as input. So, I guess once I finished that, which might take some time, I'll can modify the convolution layer to just use a matrix as input instead of a 3rd-order tensor.

I think we should remove the cube support from the train split method, and use a matrix to store images.

rcurtin commented:
We could also provide some functionality to convert from a cube representation of an image to a flattened matrix.  It's a lot easier in the internal code to support only one type of object, especially because in this case the slices and columns would have to be handled differently.

I suppose we could add something like `data::FlattenCube(const arma::Cube<eT>& cube, arma::mat& matrix);`, but honestly I'm not sure that many people are loading their images as cubes (does Armadillo have built-in support for that?  I know `data::Load()` won't do it).

stereomatchingkiss commented:
>We could also provide some functionality to convert from a cube representation of an image to a flattened matrix.

I think this could be omitted, use arma::mat to present an image is more than enough and it should be easier to do the vectorization.

Since both of you agree on arma::cube should be removed, I will remove the api from the codes.

rcurtin commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,155 @@
+    testLabel.set_size(testSize);
+
+    using Col = arma::Col<size_t>;
+    arma::arma_rng::set_seed(seed);
```
I don't think there's any need for this, we should just assume that the user has called `math::RandomSeed()` appropriately.

rcurtin commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,155 @@
+
+  /**
+   *Split training data and test data, please define
+   *ARMA_USE_CXX11 to enable move of c++11
```
mlpack requires C++11 so I think we can assume that `ARMA_USE_CXX11` is enabled without any issues.

rcurtin commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,155 @@
+
+private:  
+  arma::arma_rng::seed_type seed;
+  size_t slice;
```
I guess we can remove this member and related functions if we are not aiming to support cubes?

rcurtin commented on src/mlpack/tests/split_data_test.cpp:
```diff
@@ -0,0 +1,51 @@
+            BOOST_REQUIRE_CLOSE(lhsCol(j), rhsCol(j), 1e-5);
+        }
+    }
+}
```
Do you mind adding a comment saying what this function does?  It might also be worth noting that this function depends on the labels being equivalent to the original point index (the trick you use in line 36/37).

rcurtin commented:
This looks good to me, but with the simpler API that does not accept cubes, do you think it is worth having a class for this, or would it be more reasonable to just have a function `TrainTestSplit()` which takes the data and the percentage for the test set?  I think that since `TrainTestSplit` only holds one member it would be easier to just make it a standalone function.  You could also add the overload that returns the `std::tuple<...>` too.

rcurtin commented:
Also sorry for the slow response on this one... :(

stereomatchingkiss commented:
>mlpack requires C++11 so I think we can assume that ARMA_USE_CXX11 is enabled without any issues.

I am not sure about other compiler, but this do not work on vc2015, you have to specify it else armadillo do not support move.

>Also sorry for the slow response on this one... :(

Don't mind, we all have our daily jobs to be done, recently I am wandering around the la la land of CBIR and web apps too.

rcurtin commented:
Hmm, do you think that we should modify `prereqs.hpp` or something to define `ARMA_USE_CXX11` when using Visual Studio 2015?  I am fine with this if everything still compiles correctly.  Also maybe I am missing it but I don't see where `std::move()` is used, so maybe the comment is no longer necessary?  Not certain on that.

> Don't mind, we all have our daily jobs to be done, recently I am wandering around the la la land of CBIR and web apps too.

The CBIR part sounds cool to me! :)

stereomatchingkiss commented:
>Hmm, do you think that we should modify prereqs.hpp or something to define ARMA_USE_CXX11 when using Visual Studio 2015?

I think this is a good idea if everything still compile correctly

> Also maybe I am missing it but I don't see where std::move() is used, so maybe the comment is no longer necessary?

I do not use std::move to move them explicitly because this may generate less effective binary, compiler will try to do RVO if they could, if they could not, they would move the variable back, compiler is smart enough to find out they are local variable only needed by that function.

rcurtin commented:
This looks good to me; when it's merged I'll probably rewrite the documentation a bit to clarify everything if that's okay.  @zoq: any comments before merge?

zoq commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,108 @@
+                    double const testRatio)
+{
+  size_t const testSize =
+      static_cast<size_t>(input.n_cols * testRatio);
```
I'm not sure why we need an static_cast here. I guess it's because we can check the type at compile-time?

stereomatchingkiss replied:
Because the compiler may give you type cast warning if you do not add static_cast

zoq commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,108 @@
+{
+  size_t const testSize =
+      static_cast<size_t>(input.n_cols * testRatio);
+  size_t const trainSize = input.n_cols - testSize;
```
Sorry to be picky, but maybe we can agree on a style? Do we use const as suffix or prefix, or are we going to accept both styles?

rcurtin replied:
I'd agree with this, I'd personally prefer prefix const since it's what's used everywhere.

stereomatchingkiss replied:
Ok, than I would make it as prefix.

zoq commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,108 @@
+  Col const sequence = arma::linspace<Col>(0, input.n_cols - 1,
+                                           input.n_cols);
+  arma::Col<size_t> const order = arma::shuffle(sequence);
+
```
Okay, another highly biased and all in all picky comment from my side. I would combine line 53-56 into a single line. We don't need to store ```sequence``` right?

stereomatchingkiss replied:
You are correct

zoq commented on src/mlpack/core/util/split_data.hpp:
```diff
@@ -0,0 +1,108 @@
+#include <random>
+#include <tuple>
+#include <vector>
+
```
It looks like, since you striped the code, we can get rid of almost all of the header files except for the tuple header?

stereomatchingkiss replied:
Thank you very much, I forgot to remove those useless headers.

rcurtin commented:
I made some simple changes, let me know if you think any of the changes should be reverted.  Mostly I just made a few style fixes, moved it to `mlpack::data` from `mlpack::util`, and added another test.

213a04d Add another test, and check that we aren't duplicating points.
35149da Style fixes.
410bed0 Move SplitData to data namespace.
c20195c Move <tuple> to standard include.
